crates/bevy_ecs/src/schedule/ambiguity_detection.rs

```diff
@@ -0,0 +1,881 @@
+                        &self,
+                        &world,
+                    )
+                })
+                .collect();
+
+        let mut ambiguities = Vec::default();
+        ambiguities.extend(at_start);
+        ambiguities.extend(parallel);
+        ambiguities.extend(before_commands);
+        ambiguities.extend(at_end);
+
+        ambiguities
+    }
+
+    /// Returns the number of system order ambiguities between systems in this stage
+    pub fn n_ambiguities(
+        &mut self,
+        world: &mut World,
+        report_level: ReportExecutionOrderAmbiguities,
+    ) -> usize {
+        self.ambiguities(world, report_level).len()
+    }
+
+    /// Reports all execution order ambiguities between systems
+    pub fn report_ambiguities(
+        &mut self,
+        world: &mut World,
+        report_level: ReportExecutionOrderAmbiguities,
+    ) {
+        let ambiguities = self.ambiguities(world, report_level);
+        let unresolved_count = ambiguities.len();
+
+        if unresolved_count > 0 {
+            // Grammar
+            if unresolved_count == 1 {
+                println!("One of your stages contains 1 pair of systems with unknown order and conflicting data access. \n\
+				You may want to add `.before()` or `.after()` ordering constraints between some of these systems to prevent bugs.\n");
+            } else {
+                println!("One of your stages contains {unresolved_count} pairs of systems with unknown order and conflicting data access. \n\
+				You may want to add `.before()` or `.after()` ordering constraints between some of these systems to prevent bugs.\n");
+            }
+
+            if report_level == ReportExecutionOrderAmbiguities::Minimal {
+                println!("Set the level of the `ReportExecutionOrderAmbiguities` resource to `AmbiguityReportLevel::Verbose` for more details.");
+            } else {
+                for (i, ambiguity) in ambiguities.iter().enumerate() {
+                    let ambiguity_number = i + 1;
+                    // The path name is often just noise, and this gets us consistency with `conflicts`'s formatting
+                    let system_a_name = format_type_name(ambiguity.system_names.0.as_str());
+                    let system_b_name = format_type_name(ambiguity.system_names.1.as_str());
+                    let conflicts: Vec<String> = ambiguity
+                        .conflicts
+                        .iter()
+                        .map(|name| format_type_name(name.as_str()))
+                        .collect();
+
+                    println!("{ambiguity_number:?}. `{system_a_name}` conflicts with `{system_b_name}` on {conflicts:?}");
+                }
+                // Print an empty line to space out multiple stages nicely
+                println!("");
+            }
+        }
+    }
+}
+
+/// Collapses a name returned by std::any::type_name to remove its module path
+fn format_type_name(raw_name: &str) -> String {
```

Comment:
This could probably be done with a single gnarly regex, but I'm not actually sure that would be more readable or maintainable.

Reply:
There's a function `TypeRegistration::get_short_name` that does something very similar to this function.  Not sure if it matches exactly.

Reply:
This is an exact match, although the implementation had a few bugs in it. I've refactored this out into bevy_utils and reused it.

Thanks!